### PR TITLE
defaulting webhook for clusterbootstrap cr

### LINKED ISSUE #1916: Defaulting webhook for partial ClusterBootstrap
Use case: CLI/Client wants to customize ClusterBootstrap by specific customized CNI selection or other config. Requiring a client to clone the full ClusterBootstrapTemplate puts lot of burden on ensuring all providers are cloned properly with their configs.

Proposal: Defaulting webhook for ClusterBootstrap that fills out missing information from a ClusterBootstrapTemplate based on the presence of an annotation.

The annotation could take the form
"tkg.tanzu.vmware.com/clone-missing-from": "run.tanzu.vmware.com/ClusterBootstrapTemplate/v1.23.3---vmware.1-tkg.1"

Value is string representation of corev1.TypedLocalObjectReference

We could also set a namespace in the annotation to make it explicit.

Done criteria:
Defaulting webhook fills in all remaining fields when creating a ClusterBootstrap object that has an annotation

### PR DESCRIPTION
**Pull request by danniel1205**
### What this PR does / why we need it

Use cases:
* Some fields of ClusterBootstrap are optional per API definition, but they are required by internal logic to make things work properly. We need to add defaults to those fields which are not specified.
* When users(CLI or Controller) create a ClusterBootstrap CR, a fully qualified ClusterBootstrapPackage RefName`<short-name>.<domain>.<version>` is required, however the users might not know the version. We need to complete the RefName by fetching the version from TKR.

This PR adds a defaulting logic in ClusterBootstrap webhook to support above two use cases. And a predefined annotation of ClusterBootstrap indicates which TanzuKubernetesRelease we need to get the default values from.

### Which issue(s) this PR fixes

Fixes #1916 

### Describe testing done for PR

- Local unit tests passed
- Local integ tests passed

### Release note

```release-note
package-based-lcm:
Allow adding tkg.tanzu.vmware.com/add-missing-fields-from-tkr=<tkr-name> in the annotation of ClusterBootstrap to enable the defaulting webhook logic which adds defaults to the missing spec fields and complete the partially filled ClusterBootstrapPackage RefName
```

### PR Checklist

- [ ] Squash the commits into one or a small number of logical commits
      
- [ ] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [ ] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go [hunk 1]
```diff
@@ -55,11 +57,68 @@ func NewHelper(ctx context.Context, k8sClient client.Client, aggregateAPIResourc
...
+					// TODO: Handle the primitive pointer, e.g., *int. Ideally we need to make a copy of the pointer
+					// instead of reassigning directly. It is a shallow copy with current approach, copyFrom and destination
+					// share the same underlying data. We are good at the moment because this function is internally
+					// used for handling ClusterBootstrapTemplateSpec which does not have any primitive pointers in its
+					// API definition.
+					destination[keyInFrom] = valueInFrom
+				}
+			}
+		} else {
+			// If keyInFrom exists in addMissingTo, recursively look inside the nested fields.
```
Comment (maralavi) [inline addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go:112]: ```addMissingTo ``` ---> ```destination```

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone_test.go [hunk 2]
```diff
@@ -299,8 +301,198 @@ var _ = Describe("Clusterbootstrap", func() {
...
+			Expect(clusterBootstrap.Spec.CPI.RefName).To(Equal("fake-cpi"))
+			Expect(clusterBootstrap.Spec.CPI.ValuesFrom.Inline).To(BeNil())
+			Expect(clusterBootstrap.Spec.CPI.ValuesFrom.SecretRef).To(BeEmpty())
+			Expect(clusterBootstrap.Spec.CPI.ValuesFrom.ProviderRef.Kind).To(Equal("secret"))
+			Expect(clusterBootstrap.Spec.CPI.ValuesFrom.ProviderRef.Name).To(Equal("fake-secret"))
+			// The partial filled refName should be updated
+			Expect(clusterBootstrap.Spec.CNI.RefName).To(Equal("calico.tanzu.vmware.com.3.22.1+vmware.1-tkg.1-zshippable"))
+			Expect(clusterBootstrap.Spec.CNI.ValuesFrom.Inline["foo"]).To(Equal("bar"))
+		})
+	})
```
Comment (vijaykatam) [inline addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone_test.go:454]: Should we add a test for merge of additionalPackages? Request has one additionalPackage and the rest are added from CBT? We should ensure that more than 1 package with same refName doesn't exist. Example user provides fluentbit.tanzu.vmware.com.3.1 whereas CBT has fluentbit.tanzu.vmware.com.2.1

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 3]
```diff
@@ -82,7 +83,58 @@ var _ webhook.CustomValidator = &ClusterBootstrap{}
...
+		return err
+	}
+	if err := helper.AddMissingSpecFieldsFromTemplate(clusterBootstrapTemplate, clusterBootstrap); err != nil {
+		clusterbootstraplog.Error(err, fmt.Sprintf("unable to add defaults to the missing fields of ClusterBootstrap %s/%s",
+			clusterBootstrap.Namespace, clusterBootstrap.Name))
+		return err
+	}
+
+	// Get TanzuKubernetesRelease and attempt to complete the partial filled ClusterBootstrapPackage RefName
+	tkr := &runv1alpha3.TanzuKubernetesRelease{}
```
Comment (vijaykatam) [inline addons/webhooks/clusterbootstrap_webhook.go:128]: Lets move this up before trying to add defaults so that we can catch not exists error sooner.

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 4]
```diff
@@ -82,7 +83,58 @@ var _ webhook.CustomValidator = &ClusterBootstrap{}
...
+	if err := wh.Client.Get(ctx, client.ObjectKey{Name: tkrName}, tkr); err != nil {
+		clusterbootstraplog.Error(err, fmt.Sprintf("unable to get the TanzuKubernetesRelease %s", tkrName))
+		return err
+	}
+	if err := helper.CompleteCBPackageRefNamesFromTKR(tkr, clusterBootstrap); err != nil {
```
Comment (vijaykatam) [inline addons/webhooks/clusterbootstrap_webhook.go:133]: Should we derive package names first and then do defaulting? It will help with scenarios where we don't want to end up with the same package twice in additionalPackages.

File: addons/webhooks/clusterbootstrap_webhook_test.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+					},
+				},
+				Spec: &runv1alpha3.ClusterBootstrapTemplateSpec{},
+			}
+			err := k8sClient.Create(ctx, clusterBootstrap)
+			Expect(err).To(HaveOccurred())
+			Expect(apierrors.IsNotFound(err)).To(BeTrue())
+		})
+
+		// TODO: Add more tests to verify the CompleteCBPackageRefNamesFromTKR() logic.
```
Comment (vijaykatam) [inline addons/webhooks/clusterbootstrap_webhook_test.go:142]: Could you file an issue?

**Review by vijaykatam — COMMENTED** [review]
Changes look good for the most part. Any chance you looked at https://github.com/kubernetes-sigs/structured-merge-diff if there are reusable helpers that could be used.

File: addons/webhooks/clusterbootstrap_webhook_test.go (hunk 5 above)
Comment (danniel1205) [reply addons/webhooks/clusterbootstrap_webhook_test.go]: Done. I made some changes to TKR CRD and fixed the issue.

File: addons/webhooks/clusterbootstrap_webhook.go (hunk 4 above)
Comment (danniel1205) [reply addons/webhooks/clusterbootstrap_webhook.go:129]: Done, changed the order of CompleteCBPackageRefNamesFromTKR() and AddMissingSpecFieldsFromTemplate()

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone_test.go (hunk 2 above)
Comment (danniel1205) [reply addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone_test.go:504]: Done, added more tests with non-empty `AdditionalPackages`

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go [hunk 6]
```diff
@@ -55,11 +58,111 @@ func NewHelper(ctx context.Context, k8sClient client.Client, aggregateAPIResourc
...
+		}
+	}
+	return nil
+}
+
+func getFullyQualifiedCBPackageRefName(tkr *runtanzuv1alpha3.TanzuKubernetesRelease, prefix string) (string, error) {
+	var fullyQualifiedCBPackageRefName string
+	for _, tkrBootstrapPackage := range tkr.Spec.BootstrapPackages {
+		if strings.HasPrefix(tkrBootstrapPackage.Name, prefix) {
+			fullyQualifiedCBPackageRefName = tkrBootstrapPackage.Name
```
[the remark this answers is not in the document]
Comment (danniel1205) [reply addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go]: yea, it's a good idea to prevent a too broad match.

**Comment by danniel1205** [comment]
> Changes look good for the most part. Any chance you looked at https://github.com/kubernetes-sigs/structured-merge-diff if there are reusable helpers that could be used.

Thanks for pointing it out. Yea, looked into it and found some of our cases could not be easily addressed by it. So I did not reuse it.

File: addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go [hunk 7]
```diff
@@ -55,11 +58,127 @@ func NewHelper(ctx context.Context, k8sClient client.Client, aggregateAPIResourc
...
+func getFullyQualifiedCBPackageRefName(tkr *runtanzuv1alpha3.TanzuKubernetesRelease, prefix string) (string, error) {
+	var fullyQualifiedCBPackageRefName string
+	var found bool
+	for _, tkrBootstrapPackage := range tkr.Spec.BootstrapPackages {
+		if strings.HasPrefix(tkrBootstrapPackage.Name, prefix) {
+			if found {
+				// It is possible that current prefix matches multiple bootstrapPackages. For example, c* might match
+				// capabilities bootstrapPackage and calico bootstrapPackage. We would like to return error in this case
+				// to let caller revise the prefix in order to narrow down to exactly one match.
+				return fullyQualifiedCBPackageRefName, fmt.Errorf("multiple bootstrapPackage names matche the prefix %s within the TanzuKubernetesRelease %s", prefix, tkr.Name)
```
Comment (vijaykatam) [inline addons/pkg/util/clusterbootstrapclone/clusterbootstrapclone.go:99]: nit: matche -> matches

**Review by vijaykatam — APPROVED** [review]
Minor nit. LGTM! Thanks for the changes.
